### ISSUE TITLE
Bug: GibbsComputedStructureEntry returning incorrect energies, running slowly

### ISSUE THREAD

**Issue by mattmcdermott** [materialsproject/pymatgen#2053]
**Describe the bug**
`GibbsComputedStructureEntry` is now returning incorrect formation energies and running very slowly, I think due to refactoring in #2014 (tagging @CompRhys for reference, but I should be able to address).

**To Reproduce**
```python
from pymatgen.analysis.phase_diagram import PhaseDiagram
from pymatgen import MPRester
from pymatgen.entries.computed_entries import GibbsComputedStructureEntry

with MPRester() as mpr:
    all_entries = mpr.get_entries_in_chemsys("C-O", inc_structure="final")

pd_mp = PhaseDiagram(all_entries)

temp = 900
gibbs_entries = GibbsComputedStructureEntry.from_pd(pd_mp,temp)
pd_gibbs = PhaseDiagram(gibbs_entries)

for e in sorted(list(pd_gibbs.stable_entries),key=lambda x: x.composition.reduced_formula):
    print(e.composition.reduced_formula, pd_gibbs.get_form_energy_per_atom(e))
```

**Expected behavior**
The formation energy of CO2 should be around -1.367 eV/atom at T=900 K, according to NIST-JANAF data. Instead, the energy output currently shows -3.147 eV/atom.

**Desktop (please complete the following information):**
 - OS: Mac

**Additional context**
I will be making a PR to address ASAP -- will also try to add more tests to catch bugs like this in the future.

**Comment by CompRhys** [comment]
The phase diagram parts running more slowly will be due to the equality now checking the structure in getting stable_entries. It might be worth removing the structure check from the `__eq__` if the slow down in prohibitive?  

I can't think of any edge cases with the gibbs energy correction I went through all the edge cases I could think of.

**Comment by mkhorton** [comment]
Thanks for making a PR @mattmcdermott -- could you add an explicit test for the CO2 data too? Just so we can catch a bug like this in future.

Re. equality checking, open to suggestions on the cleanest way to do that.

**Comment by mattmcdermott** [comment]
@CompRhys the issue with the refactoring is double-counting the energy for experimental entries. i.e. if the composition appears in one of the expt. tables, that expt Gibbs Free Energy is added as a `gibbs_correction` to the formation enthalpy, when it should really just be the final energy!

**Comment by CompRhys** [comment]
@mkhorton cleanest way would probably be to replace
https://github.com/materialsproject/pymatgen/blob/7d2d2001e4b9fb38bf2389a09ac0cfa3030b64ef/pymatgen/entries/__init__.py#L163
with 
```python
       return all(_is_robust_eq(other_dict[k], v) for k, v in self_dict.items() if k != "structure")
```

Although it doesn't show any speed impact for me in the above code snippet with both checking the struct and not checking it taking ~ 1s but I would guess the slow down was observed in a larger system.

**Comment by mattmcdermott** [comment]
@CompRhys I did some line profiling, and interestingly the vast majority (99%) of compute time is spent just on the `as_dict() `methods for the Entry objects. The line you referenced with `_is_robust_eq` only takes about 1% of the time.

Not quite sure how to get around that yet... definitely a drawback of computing gigantic dictionaries. Any thoughts @mkhorton ?

**Comment by CompRhys** [comment]
Using as_dict was to keep the code clean as the alternative was to have lots of custom `__eq__`. Inside the calls to as_dict the slow part will be getting the dict of the structure right?

Maybe it would be better to make `__eq__` and abstract method to avoid as_dict or could memoization of as_dict help?

**Comment by mkhorton** [comment]
> Not quite sure how to get around that yet... definitely a drawback of computing gigantic dictionaries. Any thoughts @mkhorton ?

Yes, I proposed the as_dict() as a way to get approx_equal behavior between disparate classes. Given that you're encountering performance issues, I'm inclined to say this was the wrong call and we should do something else.

**Comment by mattmcdermott** [comment]
@CompRhys The memoization idea actually does work pretty well. I saw a 6x speed-up by adding `functools.lru_cache` to the `ComputedStructureEntry.as_dict()` method. I then saw another 10x speedup by removing the requirement of computing `Structure.as_dict()`. So altogether, about a 60x speedup to cache the result and not do structure dict-checking. 

@mkhorton Is there anything wrong with doing caching for the as_dict() method? Since I'd also like to have the option to remove the structure in equality checking, maybe the better fix is to just implement a different __eq__ method for ComputedStructureEntry alone?

**Comment by mkhorton** [comment]
@mattmcdermott nothing wrong with caching, it's just a hack (caching) on top of a hack (using as_dict). But it is a fairly minimal one. To rephrase; I'm ok doing the caching, but would prefer a cleaner solution if one exists.

**Comment by CompRhys** [comment]
Caching via lru_cache might run into problems as the entries are mutable and could be changed without the cache being cleared

I think custom `__eq__` for all `EntryLike` is probably the best option but it won't be particularly clean.

**Comment by shyamd** [comment]
Honestly, if classes subclassing `Entry` are meant to be hashable, then every subclass should have to define a `__hash__` and `__eq__`. The big thing is to use a combination of `super()` functions and then check what the new subclass added in.

**Comment by CompRhys** [comment]
Another hacky speed up would be to adjust the verbosity of the structure `as_dict` call to be zero rather than default. Reading the code looking for bottlenecks it seems like this could help with the current setup but not sure what it would do practically and haven't tried it yet. It would have to be done in the `_is_dict_eq` so might require adding a `**kwargs` to a lot of `as_dict` so they don't break when given a verbosity argument.

**Comment by mattmcdermott** [comment]
I did implement a solution like that, but I think it's too hacky. @mkhorton and I are working on a fix

**Comment by mkhorton** [comment]
Closing the issue since it was fixed in #2058

**Comment by mkhorton** [comment]
I just wanted to update this thread to say that we merged in a change that makes `Entry` immutable (immutability not enforced, but there is no longer a way to mutate Entry in-place). We also swapped out the `__eq__` for a faster version to correct my mis-step of suggesting the dictionary comparisons; this includes the `entry_id` comparison as a quick way to exit the equality check. I suspect that it can still be improved further but hopefully this will address performance issues for now.
